Fix SendEmailAction's check to support multiple emails per subscriber [MAILPOET-6155]

costasovo opened the pull request:
## Description

It is possible that one email (e.g., purchase in category) is sent multiple times to the same subscriber. Due to an issue in a query that checks the status of sending, some sent email actions were marked as failed even though the email was successfully sent.

This PR fixes the query that was causing the issue.

## Code review notes

AutomationEmailScheduler::getScheduledTaskSubscriber was selecting the task based on subscriber and newsletter. In the case of multiple emails sent to one subscriber, the method failed to pick ScheduledTaskSubsrciberEntity because the query was fetching multiple results, but getOneOrNullResult expects only one result.

This commit fixes it by adding additional filtering by $runId to get the ScheduledTaskSubsriberEntity associated with the correct run.

I did the filtering in PHP because an alternative would be using LIKE %% in the query. The meta column is text. 

## QA notes

#### Steps to replication:

1. Create an automation for example Purchased in a category
3. Make sure that option Run this automation only once per subscriber is disabled (It should be by default)
5. Purchase a product from the category with the same subscriber at least twice
7. Make order as complete
9. Observe the issue on the automation analytics → tab Subscribers

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-6155]

## After-merge notes

_N/A_

## Tasks

- [x] 🚫 I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] 🚫 I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-6155]: https://mailpoet.atlassian.net/browse/MAILPOET-6155?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

costasovo committed:
Fix send action checkSendingStatus to support multiple emails per subscriber.

It is possible that one email (e.g., purchase in category) is sent multiple times
to the same subscriber.

AutomationEmailScheduler::getScheduledTaskSubscriber was selecting the task based on subscriber and newsletter.
In the case of multiple emails sent to one subscriber, the method failed to pick ScheduledTaskSubsrciberEntity because
the query was fetching multiple results, but getOneOrNullResult expects only one result.

This commit fixes it by adding additional filtering by $runId to get the ScheduledTaskSubsriberEntity associated
with the correct run.

I did the filtering in PHP because an alternative would be using LIKE %% in the query. The meta column is text.
[MAILPOET-6155]

NeosinneR reviewed (dismissed):
LGTM!

JanJakes commented on mailpoet/lib/Newsletter/Scheduler/AutomationEmailScheduler.php:
```diff
@@ -75,7 +75,19 @@ public function getScheduledTaskSubscriber(NewsletterEntity $email, SubscriberEn
+        $result = $scheduledTaskSubscriber;
+        break;
+      }
+    }
```
@costasovo Maybe I'm overthinking this, but what if there are many of such records? Won't that be suboptimal in terms of always fetching all of them? I don't have a specific scenario, but an automation can be anything... I could imagine that you send an email after every order or something, and then, if a subscriber reaches hundreds of orders, we could run into performance issues or memory errors here.

I know that addressing this properly is not easy given the `meta` field usage (would need a more complex query matching the run ID on the DBAL level).

But I have another idea — what if we limit the scheduled tasks so that `$automationRun->getCreatedAt()` is `less or equal` than scheduled task's (or sending queue's) `createdAt`?

costasovo replied:
@JanJakes Thanks for noticing the potential performance issue. I'll update the PR based on your suggestions.

costasovo committed:
Limit scheduled tasks fetching by automation run createdAt

This is a performance optimization to avoid fetching too many scheduled task rows.
Run is always created before the scheduled task which is created in send action step.
[MAILPOET-6155]

costasovo committed:
Improve TimestampListener to use the current time for now instead of "cached" value

Using now "cached" on the TimestampListener instance was causing us to set createdAt and updatedAt slightly in the past.
It was potentially problematic in tests where the TimestampListener lives a long time or in long requests, such as a cron request.

[MAILPOET-6155]

costasovo committed:
Improve check for recently scheduled tasks for sending action.

[MAILPOET-6155]

costasovo commented:
@JanJakes I added three more commits. They improve the performance of the method that searches for proper scheduled task and fix an issue with TimestampListener I ran into.

Aschepikov commented:
HI, @costasovo I created automation to reproduce the issue and checked it before the fix and after installing. Before I had an issue as was mentioned here 
![image](https://github.com/user-attachments/assets/c61e57f4-3813-49cc-8abe-ed6add3f8edd)
 and after I had another issue
![image](https://github.com/user-attachments/assets/e243d00b-43d5-433c-bf88-b48992da62cc)
With the new automation, the situation is the same everytime failed on this
![image](https://github.com/user-attachments/assets/24bb3ea5-7ddd-472b-a2e0-8d34fb3d2c6a)

Aschepikov commented:
15 mins later I received 2 emails, 1 from auto 1 and 1 from auto 2. But with the fix installed, there should be 3 emails from auto 1 and 2 from auto 2. Status is not renewed
![image](https://github.com/user-attachments/assets/b7d5bfa7-d16a-46e5-a3d4-b92d2b2df862)

costasovo commented:
@Aschepikov This is probably related to the last changes I made. I got similar errors in tests, and I thought I understood and fixed those. But apparently not :( What timezone does your site use?

costasovo committed:
Address offset issue in time comparison in AutomationEmailScheduler

[MAILPOET-6155]

costasovo commented:
@Aschepikov I was able to replicate the issue with the timezone set to Denver. Working on a fix.

costasovo commented:
@JanJakes I needed to add one more commit. I originally misunderstood how offset on DB connection works.
The problem is that AutomationRun is handled via WPDB, which stores and returns UTC values.
The PDO connection uses gmt_offset and it stores UTC, but returns and also expects values as UTC+offset.

The `created_at` comparison was failing for negative offset because the PDO connection expects UTC+offset as the parameter value, but the value from AutomationRun is UTC because it was fetched via WPDB. The exact fix would be adding offset to the AutomationRun's created_at before passing it as a parameter. The problem will be solved with the switch to WPDB. Here, we need to reduce the amount of fetched rows because of performance, so I went with a simple solution subtracting one day.

JanJakes reviewed (approved):
Awesome, thanks for tackling this!

Aschepikov reviewed (approved):
GJ!
